### PR TITLE
[Serve] Document how to stream outputs from `@serve.batch`

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

#35729 and  #36136 let users stream outputs from `@serve.batch`-decorated functions. This change adds documentations explaining how to use this feature.

Updated doc link: https://anyscale-ray--36263.com.readthedocs.build/en/36263/serve/performance.html#streaming-batched-requests

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [X] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
Looks good, but we should also add a more practical example with text generation. Are you planning to do that separately?

**Comment by shrekris-anyscale** [comment]
Yeah, I think it would be better to do it separately. I'm planning to add a doc to the `Examples` section.

File: doc/source/serve/doc_code/batching_guide.py
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:53]: missing `await`

File: doc/source/serve/doc_code/batching_guide.py
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:63]: please add tests here before merging

File: doc/source/serve/doc_code/batching_guide.py
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:88]: missing await

File: doc/source/serve/doc_code/batching_guide.py
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:87]:
```suggestion
            yield [
                i for i, requested_max in enumerate(max_list) if i <= requested_max else StopIteration
            ]
```
might be easier to read? not sure

File: doc/source/serve/performance.md
Comment (edoakes) [inline doc/source/serve/performance.md:126]:
```suggestion
Use an async generator to stream the outputs from your batched requests. Let's convert the `StreamingResponder` class to accept a batch.
```

File: doc/source/serve/performance.md
Comment (edoakes) [inline doc/source/serve/performance.md:135]:
```suggestion
Decorate async generator functions with the {mod}`ray.serve.batch` decorator. Similar to non-streaming methods, the function takes in a `List` of inputs and in each iteration it `yield`s an iterable of outputs with the same length as the input batch size.
```

**Review by edoakes — APPROVED** [review]
LGTM pending nits

File: doc/source/serve/doc_code/batching_guide.py
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: Thanks, I added the `await`.

File: doc/source/serve/doc_code/batching_guide.py
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: I added some tests.

File: doc/source/serve/doc_code/batching_guide.py
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: Thanks, I added the `await`.

File: doc/source/serve/doc_code/batching_guide.py
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: This feels harder to read in my opinion. I'd prefer to stick with the explicit `if`-`else`.

**Review by angelinalg — APPROVED** [review]

**Pull request merged**
